### ISSUE TITLE
wgsl: Missing specification on modf behavior for negatives.

### ISSUE THREAD

**Issue by Axiverse** [gpuweb/gpuweb#2468]
Per https://stackoverflow.com/questions/7610631/glsl-mod-vs-hlsl-fmod, glsl and hlsl have different behaviors on their mod operators. While wgsl defines the modf method, it does not specify the behavior for negative numbers.

1. `modf(-0.4) = {whole: -1, fract: 0.6}`
1. `modf(-0.4) = {whole: 0, fract: -0.4}`

Is this something that can be standardized, or is the behavior with negative number undefined? Either way, I believe it should be described in the spec.

**Comment by stegu** [comment]
This is a major issue for modulo arithmetic. Modulo fields are not always powers of two that can be implemented with bitmasks, and it's often convenient to be able to do modulo operations on floats. Periodic noise, which is what I am doing right now in WGSL, is one major use case here.
The operator "%" is, unfortunately, wrong for that purpose. It's a "remainder" operator, not a proper modulo operator. A built-in mod() function that behaves like mod() in GLSL, i.e. mod(a,b) = a-floor(a/b)*b, would probably be a good idea, because it could be both accelerated and implemented with better precision as a built-in than as a user defined function, and save us the hassle of having to implement and name a different user function for every data type to perform the equivalent of GLSL mod().

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-05-03-minutes#wgsl-missing-specification-on-modf-behavior-for-negatives--issue-2468">WGSL meeting minutes 2022-05-03</a></summary>



* KG: Thought we figured this out
* MM: Feel like we should do what we did for integrals to floating point values
* BC: Spent time trying to agree on what flavour of modf we’re doing. Hadn’t resolved mismatch between int and float. Not sure if that’s what this refers too
* AB: Spec states implementation of floating point modf. Not sure what more is needed. Separate comment on issue about some people preferring GLSL behaviour which does not match HLSL or MSL. We ended up with Spirv remainder vs mod as it’s more consistent. In past talked about buitlin function to provide both behaviour but backed away previously
* KG: Leave comment we deferred to SPIR-V spec and therefore is well defined?
* AB: SPIR-V mappings were taken out of spec. Now for fp type we give the formula in the spec directly. So, I think it’s properly defined and matches, but not written exact same way, what is in the int section
* MM: **Should link to section of spec, close issue and say if behaviour is not desired to open a new issue.**


</details>

**Comment by kdashg** [comment]
However, looking for it, I can't find *any* specification of modulo, at least not by name in the WGSL spec.
@alan-baker Can you find it?

**Comment by dneto0** [comment]
> However, looking for it, I can't find _any_ specification of modulo, at least not by name in the WGSL spec. @alan-baker Can you find it?

As @alan-baker said in the meeting (notes above), previously this was defined via reference to the SPIR-V spec. But that has been removed.

See [here](https://github.com/gpuweb/gpuweb/pull/2594/files#diff-709b5c343d59ae10cbabe69f9df1eee9e2dbb6e9d848632fa94357083d6e0a34L9339), where WGSL had referenced the ModfStruct extended instruction (in https://www.khronos.org/registry/SPIR-V/specs/1.0/GLSL.std.450.html ).  There it refers to Modf, which says that for  Modf(x), "Both the result and the output parameter will have the same sign as x."


The WGSL spec does need to spell out the intended behaviour.   

For modf(x), the whole part equals   (x % 1.0), and  the fractional part is x - (x % 1.0).    The floating point % spells out the behaviour for all cases.

**Comment by dneto0** [comment]
Given the meeting notes, I think the status should be Resolved: needs specification work.

**Comment by dneto0** [comment]
From the meeting notes:

> MM: Should link to section of spec, close issue and say if behaviour is not desired to open a new issue.

PR #2835 essentially links the behaviour explicitly to the floating point % operator.

If the behaviour is not desired, open a new issue.  As Alan noted, we had considered adding more builtins to get the alternate result, but had rejected that for version 1.0 of the language.

**Comment by brent-williams** [comment]
Minor correction @dneto0 for anyone following along:

> For modf(x), the whole part equals (x % 1.0), and the fractional part is x - (x % 1.0).

I think this should read:

For modf(x), the fractional part equals (x % 1.0), and the whole part is x - (x % 1.0).

**Comment by stegu** [comment]
TL;DR: Use case?

I think this is going down the wrong path and cementing a bad choice. A function that returns (x % 1.0) as its fractional part should be named "remf", not "modf". A modulo operation has a strict mathematical definition, and the GLSL mod() function has it right.
What is the use case for a modf() function with this proposed behaviour? The use case for the alternative, where the fractional part is always positive, is to create a coordinate wrap with a result where the fractional part is a sawtooth wave that is invariant under integer shifts, i.e. does not flip the direction of its periodic ramp at the origin. I really, *really* think the proper behaviour is that if modf(x) = {f,i}, then modf(x+N) = {f, i+N} for any integer N.
My strongly held professional opinion is that modf(x) should return mod(x, 1.0) for the fractional part, with the GLSL behaviour for mod(), and floor(x) for the integer part.
An extension to a general (and useful) floating point mod with two operands and two return values would be logical as well: modf(x,p) = {mod(x,p), floor(x/p)}, where mod(x,p) = x - floor(x/p). In this case, the second return argument simply passes the internally computed floor(x/p) out for external use.
Another function named remf(x), if needed, could return (x%1.0) and trunc(x), with a possible extension to remf(x,p) = {x%p, trunc(x/p)}, although I think that function would be better named remt() rather than remf(). However, I see no obvious use for remf()/remt().
Modulo operations are central to pattern generation in computer graphics. Remainder operations are not. Like I said, what is the use case for the currently proposed remainder-like behaviour, and why is the function named modf() when it does not describe a modulo operation?
The confusion of remainder and modulo operations is a hassle in many programming languages. I think the culprit is the % operator, which puts rem() undeservedly front and center when mod() is the considerably more useful option.

**Comment by jespertheend** [comment]
I too need the glsl implementation waaay more frequently than the one currently defined in the spec.
I don't think I ever needed a function where a negative input resulted in a negative fractional output to be honest.

**Comment by jespertheend** [comment]
Which operation is more useful *is* a matter of opinion I think.
I do a lot of work in game development where I use modulo more frequently. But I can see how maybe in another field like machine learning, which I know nothing about, the remainder operation might be more useful.

What *isn't* a matter of opinion is the name that's used. Modulo and remainder should not be confused with each other, which is exactly what's happening here.
